core/drivers/crypto/crypto_api/cipher/cipher.c

```diff
@@ -0,0 +1,200 @@
+{
+	assert(ctx && ctx->ops == &cipher_ops);
+
+	return container_of(ctx, struct crypto_cipher, cipher_ctx);
+}
+
+/*
+ * Free cipher context
+ *
+ * @ctx    Reference the API context pointer
+ */
+static void cipher_free_ctx(struct crypto_cipher_ctx *ctx)
+{
+	struct crypto_cipher *cipher = to_cipher_ctx(ctx);
+
+	if (cipher->op && cipher->op->free_ctx)
+		cipher->op->free_ctx(cipher->ctx);
+
+	free(cipher);
+}
+
+/*
+ * Copy cipher context
+ *
+ * @dst_ctx  [out] Reference the API context pointer destination
+ * @src_ctx  Reference the API context pointer source
+ */
+static void cipher_copy_state(struct crypto_cipher_ctx *dst_ctx,
+			      struct crypto_cipher_ctx *src_ctx)
+{
+	struct crypto_cipher *cipher_src = to_cipher_ctx(src_ctx);
+	struct crypto_cipher *cipher_dst = to_cipher_ctx(dst_ctx);
+
+	if (cipher_src->op && cipher_src->op->copy_state)
+		cipher_src->op->copy_state(cipher_dst->ctx, cipher_src->ctx);
+}
+
+/*
+ * Initialization of the cipher operation
+ *
+ * @ctx      Reference the API context pointer
+ * @mode     Operation mode
+ * @key1     First Key
+ * @key1_len Length of the first key
+ * @key2     Second Key
+ * @key2_len Length of the second key
+ * @iv       Initial Vector
+ * @iv_len   Length of the IV
+ */
+static TEE_Result cipher_init(struct crypto_cipher_ctx *ctx,
+			      TEE_OperationMode mode, const uint8_t *key1,
+			      size_t key1_len, const uint8_t *key2,
+			      size_t key2_len, const uint8_t *iv, size_t iv_len)
+{
+	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
+	struct crypto_cipher *cipher = to_cipher_ctx(ctx);
+	struct drvcrypt_cipher_init dinit = {};
+
+	if ((!key1 && key1_len) || (!key2 && key2_len) || (!iv && iv_len)) {
+		CRYPTO_TRACE("One of the key is not correct");
+		CRYPTO_TRACE("key1 @%p-%zu bytes", key1, key1_len);
+		CRYPTO_TRACE("key2 @%p-%zu bytes", key1, key1_len);
+		CRYPTO_TRACE("iv   @%p-%zu bytes", iv, iv_len);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (cipher->op && cipher->op->init) {
+		/* Prepare the initialization data */
+		dinit.ctx = cipher->ctx;
+		dinit.encrypt = (mode == TEE_MODE_ENCRYPT);
+		dinit.key1.data = (uint8_t *)key1;
+		dinit.key1.length = key1_len;
+		dinit.key2.data = (uint8_t *)key2;
+		dinit.key2.length = key2_len;
+		dinit.iv.data = (uint8_t *)iv;
+		dinit.iv.length = iv_len;
+
+		ret = cipher->op->init(&dinit);
+	}
+
+	CRYPTO_TRACE("cipher ret 0x%" PRIX32, ret);
+	return ret;
+}
+
+/*
+ * Update of the cipher operation
+ *
+ * @ctx        Reference the API context pointer
+ * @last_block True if last block to handle
+ * @data       Data to encrypt/decrypt
+ * @len        Length of the input data and output result
+ * @dst        [out] Result block of the operation
+ */
+static TEE_Result cipher_update(struct crypto_cipher_ctx *ctx, bool last_block,
+				const uint8_t *data, size_t len, uint8_t *dst)
+{
+	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
+	struct crypto_cipher *cipher = to_cipher_ctx(ctx);
+	struct drvcrypt_cipher_update dupdate = {};
```

Comment:
I have a question about this formatting:
What is the reason why we prefer `{ }` over `{}` ?
I forgot to ask @jenswi-linaro about this one in https://github.com/OP-TEE/optee_os/pull/3591#discussion_r378088901
If this formatting is required why is this rule not enforced by checkpatch?

Reply:
FWIW, Linux (our reference re. code formatting) uses both forms but more often with no space...
```
jerome@matebook:~/work/linux (master)$ git grep  '= {};' | wc -l
1893
jerome@matebook:~/work/linux (master)$ git grep  '= { };' | wc -l
621
```

Reply:
checkpatch is not 100%, but it takes care of a lot. We agreed on `{ }` with a space when this pattern was introduced. Skipping the space is possible, but it is a bit annoying. Who wants to annoy reviewers? ;-)

The numbers for OP-TEE are:
%git grep  '= {};' | wc -l
5
%git grep  '= { };' | wc -l
113

Reply:
Annoying reviewers is not in my interest for sure :laughing:
Internally we use clang-format for code formatting and it formats like this `{}`.
If both would be accepted, I would go for `{}` since clang-format does it too.
But no worries, I will change it to `{ }`

Reply:
@jenswi-linaro +1, in fact `{ }` with a space is even documented at https://optee.readthedocs.io/en/latest/general/coding_standards.html.